component/common/loki/client/queue_client.go

```diff
@@ -0,0 +1,601 @@
+package client
+
+import (
+	"bufio"
+	"bytes"
+	"context"
+	"errors"
+	"fmt"
+	"io"
+	"net/http"
+	"strconv"
+	"sync"
+	"time"
+
+	"github.com/go-kit/log"
+	"github.com/go-kit/log/level"
+	agentWal "github.com/grafana/agent/component/common/loki/wal"
+	"github.com/grafana/dskit/backoff"
+	"github.com/prometheus/common/config"
+	"github.com/prometheus/common/model"
+	"github.com/prometheus/prometheus/tsdb/chunks"
+	"github.com/prometheus/prometheus/tsdb/record"
+
+	"github.com/grafana/loki/pkg/ingester/wal"
+	"github.com/grafana/loki/pkg/logproto"
+	lokiutil "github.com/grafana/loki/pkg/util"
+)
+
+// StoppableWriteTo is a mixing of the WAL's WriteTo interface, that is Stoppable as well.
+type StoppableWriteTo interface {
+	agentWal.WriteTo
+	Stoppable
+	StopNow()
+}
+
+type MarkerHandler interface {
+	UpdateReceivedData(segmentId, dataCount int) // Data queued for sending
+	UpdateSentData(segmentId, dataCount int)     // Data which was sent or given up on sending
+	Stoppable
+}
+
+// queuedBatch is a batch specific to a tenant, that is considered ready to be sent.
+type queuedBatch struct {
+	TenantID string
+	Batch    *batch
+}
+
+// queue wraps a buffered channel and a routine that reads from it, sending batches of entries.
+type queue struct {
+	client *queueClient
+	q      chan queuedBatch
+	quit   chan struct{}
+	wg     sync.WaitGroup
+	logger log.Logger
+}
+
+func newQueue(client *queueClient, size int, logger log.Logger) *queue {
+	q := queue{
+		client: client,
+		q:      make(chan queuedBatch, size),
+		quit:   make(chan struct{}),
+		logger: logger,
+	}
+
+	q.wg.Add(1)
+	go q.run()
+
+	return &q
+}
+
+// enqueue is a blocking operation to add to the send queue a batch ready to be sent.
```

Comment:
I think it "may be blocking"? The channel is buffered, so it may or may not be full.

Reply:
Fixed comment, I was referring that it might block if the backing channel has no capacity.